### PR TITLE
[WIP] V3.1

### PR DESCRIPTION
**Pull request by oleorhagen**

### CODE REVIEW

**Comment by oleorhagen** [comment]
Not finished, but what is done, and should probably be looked at: 
* HeaderInfoV3
* TypeInfoV3
* WriteArtifact
* ReadArtifact
* Api refactorization into parameter structs

Ignore the cli modifications for now. Also, although there are many comits, I would recommend looking at this in one blob.

What is missing:
- header and manifest augment.
- cli (Should this be added?)

File: areader/reader.go [hunk 1]
```diff
@@ -221,14 +281,15 @@ func signatureReadAndVerify(tReader *tar.Reader, message []byte,
 	if verify == nil && signed {
 		return errors.New("reader: verify signature callback not registered")
 	} else if verify != nil {
+		fmt.Println("signatureReadAndVerify: ")
```
Comment (pasinskim) [inline areader/reader.go]: Probably you don't want to leave it here, do you?

**Comment by oleorhagen** [comment]
Mind you, most of the lines are tests though.

**Comment by oleorhagen** [comment]
@pasinskim I propose to make this into one PR for the updated:
* HeaderInfoV3
*TypeInfoV3
*WriteArtifactV3

And then make a task to rewrite the parsing of all the artifact versions.

**Comment by oleorhagen** [comment]
If anyone feels like finishing this while I'm gone, have a look at the message in the last commit.

**Pull request closed without merging**